tools/integration_tests/cloud_profiler/with_gcp_profiler_service_test.go

```diff
@@ -0,0 +1,203 @@
+					t.Logf("Read file operation completed %d times.", readFileCnt)
+				}
+			}
+		}
+	}
+}
+
+// listProfilesForServiceAndVersion queries the Cloud Profiler API for profiles
+// matching the given service name and version within the specified time window.
+// It handles pagination and basic retries for transient errors.
+// Ref: https://cloud.google.com/profiler/docs/reference/v2/rest
+func listProfilesForServiceAndVersion(
+	ctx context.Context,
+	t *testing.T, // Pass testing.T for logging within the helper
+	profilerAPIClient *gcpProfiler.Service,
+	projectID string,
+) ([]*gcpProfiler.Profile, error) {
+
+	t.Logf("Querying profiles for service [%s] version [%s]", testServiceName, testServiceVersion)
+	var profiles []*gcpProfiler.Profile
+	maxPagesToFetch := 5 // Limiting total list calls, 1000 per calls.
+	pagesFetched := 0
+	var pageToken string
+	for {
+		pagesFetched++
+		if pagesFetched > maxPagesToFetch {
+			t.Logf("Reached max pages (%d) to fetch. Stopping profile query.", maxPagesToFetch)
+			break
+		}
+
+		listCall := profilerAPIClient.Projects.Profiles.List(fmt.Sprintf("projects/%s", projectID))
+		if pageToken != "" {
+			listCall.PageToken(pageToken)
+		}
+		resp, errCall := listCall.Do()
+		if errCall != nil {
+			// Basic retry for transient errors
+			if strings.Contains(errCall.Error(), "try again") || strings.Contains(errCall.Error(), "unavailable") {
+				t.Logf("List profiles call failed, retrying once: %v", errCall)
+				time.Sleep(10 * time.Second)
+				resp, errCall = listCall.Do() // Retry the call
+			}
+			if errCall != nil {
+				return nil, fmt.Errorf("failed to list profiles from API: %w", errCall)
+			}
+		}
+
+		t.Logf("Size of the response: %d", len(resp.Profiles))
+		// Filter by service name and version on the client side.
+		for _, p := range resp.Profiles {
+			if p.Deployment != nil && p.Deployment.Target == testServiceName {
+				profileAPIVersion := ""
+				if p.Deployment.Labels != nil {
+					profileAPIVersion = p.Deployment.Labels["version"] // "version" is the label key used by the agent
+				}
+				if profileAPIVersion == testServiceVersion {
+					profiles = append(profiles, p)
+				}
+			}
+		}
+
+		if resp.NextPageToken == "" {
+			break // No more pages
+		}
+		pageToken = resp.NextPageToken
+		t.Logf("Fetching next page of profiles (token: %s)...", pageToken)
+	}
+
+	return profiles, nil
+}
+
+func TestValidateProfilerWithActualService(t *testing.T) {
+	// Setup directory and file to perform read workload.
+	randomData, err := operations.GenerateRandomData(5 * 1024 * 1024)
+	if err != nil {
+		t.Fatalf("operations.GenerateRandomData: %v", err)
+	}
+	testDirPath := path.Join(setup.MntDir(), testDirName)
+	filePath := path.Join(testDirPath, "a.txt")
+	setup.SetupTestDirectory(testDirName)
+	operations.CreateFileWithContent(filePath, 0644, string(randomData), t)
+	// Start Workload and Allow Time for Profiling
+	performRepeatedRead(t.Context(), t, filePath, 3*time.Minute)
```

Comment:
What if we remove this line? GCSFuse process still runs and we will still capture the profiles.

Reply:
Yes, this is optional. But added a workload to make it more realistic.